### PR TITLE
Adding UIA's PositionInSet and SizeOfSet properties to ToolStripMenuItemAccessibleObject

### LINKED ISSUE #3471: [Accessibility] Narrator cannot announce order information of the sub menu items in MenuStrip
* .NET Core Version: 
.Net 5.0.100-preview.7.20314.1 from Net 5.0 master branch

* Have you experienced this same bug with .NET Framework?: 
Yes

**Minimal repro:**
1. Extract WinformsFormsApp6.zip with MenuStrip and open it in VS, then build and run this application.
[WindowsFormsApp6.zip](https://github.com/dotnet/winforms/files/4803525/WindowsFormsApp6.zip)
2. Start Narrator, and focus on File menu, then expand it, go through its sub-items “New, Open, Save, Save As, Print, Print Preview, Exist” one by one

**Actual behavior:**
Narrator cannot announce order information of the sub menu items in MenuStrip:
image

**Expected behavior:**
Narrator can announce order information of the sub menu items in MenuStrip

### PR DESCRIPTION
**Pull request by v-elnovikova**
Fixes #3471

## Proposed changes

- Added UIA's PositionInSet and SizeOfSet properties to ToolStripMenuItemAccessibleObject

## Customer Impact

- Screen readers can announce menu items order information 

## Regression? 

- No

## Risk

- Minimal

### Before
3471 before

### After
3471 after

## Test methodology 

- Manually
 

## Test environment(s) 

- .NET 7.0.1
- Windows 11

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/dotnet/winforms/pull/8408)

### CODE REVIEW

File: src/System.Windows.Forms/src/System/Windows/Forms/ToolStripMenuItem.ToolStripMenuItemAccessibleObject.cs [hunk 1]
```diff
@@ -50,9 +50,71 @@ public override AccessibleStates State
                 propertyID switch
                 {
                     UiaCore.UIA.AcceleratorKeyPropertyId => _owningToolStripMenuItem.GetShortcutText(),
+                    UiaCore.UIA.PositionInSetPropertyId => GetPositionInSet(),
+                    UiaCore.UIA.SizeOfSetPropertyId => GetSizeOfSet(),
                     _ => base.GetPropertyValue(propertyID)
                 };
 
+            /// <summary>
+            /// Gets <see cref="UiaCore.UIA.PositionInSetPropertyId"/> property value.
```
Comment (Tanya-Solyanik) [inline src/System.Windows.Forms/src/System/Windows/Forms/ToolStripMenuItem.ToolStripMenuItemAccessibleObject.cs:59]:
```suggestion
            ///  Gets <see cref="UiaCore.UIA.PositionInSetPropertyId"/> property value.
```

File: src/System.Windows.Forms/src/System/Windows/Forms/ToolStripMenuItem.ToolStripMenuItemAccessibleObject.cs [hunk 2]
```diff
@@ -50,9 +50,71 @@ public override AccessibleStates State
...
+                }
+
+                foreach (ToolStripItem item in displayedItems)
+                {
+                    if (item == _owningToolStripMenuItem)
+                    {
+                        break;
+                    }
+
+                    if (item is ToolStripSeparator)
```
Comment (Tanya-Solyanik) [inline src/System.Windows.Forms/src/System/Windows/Forms/ToolStripMenuItem.ToolStripMenuItemAccessibleObject.cs:84]: Nice!

File: src/System.Windows.Forms/src/System/Windows/Forms/ToolStripMenuItem.ToolStripMenuItemAccessibleObject.cs [hunk 3]
```diff
@@ -50,9 +50,71 @@ public override AccessibleStates State
...
+                    {
+                        index--;
+                    }
+                }
+
+                return index + 1; // UIA_PositionInSet is 1-based
+            }
+
+            /// <summary>
+            /// Gets <see cref="UiaCore.UIA.SizeOfSetPropertyId"/> property value.
```
Comment (Tanya-Solyanik) [inline src/System.Windows.Forms/src/System/Windows/Forms/ToolStripMenuItem.ToolStripMenuItemAccessibleObject.cs:94]:
```suggestion
            ///  Gets <see cref="UiaCore.UIA.SizeOfSetPropertyId"/> property value.
```

**Review by Tanya-Solyanik — COMMENTED** [review]
Could you please add a unit test?

**Comment by NikitaSemenovAkvelon** [comment]
Have you considered using properties instead of methods to avoid calculation every time?

**Comment by v-elnovikova** [comment]
> Have you considered using properties instead of methods to avoid calculation every time?

Calculation here is minor and generally is not called too often.
Using properties would require tracking changes in the collection of sibling menu items, i.e., tracking the state of parent item.

File: src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/ToolStripMenuItem.ToolStripMenuItemAccessibleObjectTests.cs [hunk 4]
```diff
@@ -69,6 +69,86 @@ public void ToolStripMenuItemAccessibleObject_GetPropertyValue_ControlType_IsExp
...
+
+            Assert.Null(accessibilityObject.GetPropertyValue(UiaCore.UIA.PositionInSetPropertyId));
+        }
+
+        [WinFormsFact]
+        public void ToolStripMenuItemAccessibleObject_GetPropertyValue_PositionInSet_ReturnsExpected()
+        {
+            using MenuStrip menuStrip = new();
+
+            ToolStripMenuItem item1 = new();
```
Comment (dmitrii-drobotov) [inline src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/ToolStripMenuItem.ToolStripMenuItemAccessibleObjectTests.cs:87]:
```suggestion
            using ToolStripMenuItem item1 = new();
```
Same for `separator` and `item2` in this method and also in `ToolStripMenuItemAccessibleObject_GetPropertyValue_SizeOfSet_ReturnsExpected`.

**Review by dmitrii-drobotov — COMMENTED** [review]
Looks good!

**Review by dmitrii-drobotov — APPROVED** [review]

File: src/System.Windows.Forms/src/System/Windows/Forms/ToolStripMenuItem.ToolStripMenuItemAccessibleObject.cs [hunk 5]
```diff
@@ -50,9 +50,71 @@ public override AccessibleStates State
...
+                        break;
+                    }
+
+                    if (item is ToolStripSeparator)
+                    {
+                        index--;
+                    }
+                }
+
+                return index + 1; // UIA_PositionInSet is 1-based
```
Comment (dkazennov) [inline src/System.Windows.Forms/src/System/Windows/Forms/ToolStripMenuItem.ToolStripMenuItemAccessibleObject.cs:90]: This is a good founding.

**Review by dkazennov — APPROVED** [review]
Both fix and tests are good.

**Review by dreddy-work — APPROVED** [review]

**Pull request merged**